core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala

```diff
@@ -147,6 +148,230 @@ class ConsumerBounceTest extends IntegrationTestHarness with Logging {
+   * though offsets cannot be successfully committed.
+   */
+  @Test
+  def testCloseNoTimeoutWithClusterFailure() {
+    val numRecords = 10
+    sendRecords(numRecords)
+    val requestTimeout = 6000
+    this.consumerConfig.setProperty(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, "5000")
+    this.consumerConfig.setProperty(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG, "1000")
+    this.consumerConfig.setProperty(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG, requestTimeout.toString)
+    val consumer = createNewConsumer
+    consumer.assign(Collections.singletonList(tp))
+    receiveRecords(consumer, numRecords)
+    servers.foreach(server => killBroker(server.config.brokerId))
+    closeAndValidate(consumer, Long.MaxValue, Some(requestTimeout), Some(requestTimeout))
+  }
+
+  /**
+   * Consumer is closed during rebalance. Close should leave group and close
+   * immediately if rebalance is in progress.
+   */
+  @Test
+  def testCloseDuringRebalance() {
+    val executor = Executors.newSingleThreadExecutor
+    try {
+      closeDuringRebalance(executor, true)
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  /**
+   * Consumer closed during rebalance when brokers are not available.
+   * Close should terminate immediately since coordinator not known.
+   */
+  @Test
+  def testCloseDuringRebalanceBrokersUnavailable() {
+    val executor = Executors.newSingleThreadExecutor
+    try {
+      closeDuringRebalance(executor, true)
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  private def closeDuringRebalance(executor: ExecutorService, brokersAvailableDuringClose: Boolean) {
+    val topic = "closetest"
+    TestUtils.createTopic(this.zkUtils, topic, 10, serverCount, this.servers)
+    this.consumerConfig.setProperty(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, "60000")
+    this.consumerConfig.setProperty(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG, "1000")
+
+    def subscribeAndPoll(consumer: KafkaConsumer[Array[Byte], Array[Byte]]): Future[Any] = {
+      executor.submit(new Runnable() {
+        def run() {
+          consumer.subscribe(Collections.singletonList(topic))
+          consumer.poll(0)
+        }}, 0)
+    }
+
+    def waitForRebalance(consumer: KafkaConsumer[Array[Byte], Array[Byte]], future: Future[Any], timeoutMs: Long) {
+      val startMs1 = System.currentTimeMillis
+      while (System.currentTimeMillis < startMs1 + timeoutMs && !future.isDone)
+          consumer.poll(1000)
+      assertTrue("Rebalance did not complete in time", future.isDone)
+    }
+
+    val consumer1 = createNewConsumer
+    val future1 = subscribeAndPoll(consumer1)
+    future1.get(2000, TimeUnit.MILLISECONDS)
+
+    val consumer2 = createNewConsumer
+    val future2 = subscribeAndPoll(consumer2)
+    waitForRebalance(consumer1, future2, 2000)
+
+    val consumer3 = createNewConsumer
+    val future3 = subscribeAndPoll(consumer3)
+    // Wait for consumer3 to poll and trigger rebalance
+    Thread.sleep(2000)
```

Comment:
Can we use MockTime here instead. Would be preferable to slowing the test down by 2 whole seconds.